File: DataStructures/LinearDataStructures/C++/queue.cpp
#include<iostream>

namespace data_structures {
#include "linked_list.cpp"

template<class T>
class Queue {
public:

    Queue();

    void enqueue(T);

    T dequeue();

    T peek();

    unsigned int getSize();
    bool isEmpty();
private:
    data_structures::LinkedList<T> *list;
};


template<class T>
Queue<T>::Queue() {
    this->list = new data_structures::LinkedList<T>();
}

template<class T>
void Queue<T>::enqueue(T value) {
    this->list->addTail(value);
}

template<class T>
T Queue<T>::dequeue() {
    T value = this->list->getHead();
    this->list->removeHead();
    return value;
}

template<class T>
T Queue<T>::peek() {
    return this->list->getHead();
}

template<class T>
unsigned int Queue<T>::getSize() {
    return this->list->getSize();
}

template<class T>
bool Queue<T>::isEmpty() {
    return (this->getSize()) == 0;
}

}

File: DataStructures/LinearDataStructures/Queue.cs
namespace LinearDataStructures
{
    using System;
    using System.Collections;
    using System.Collections.Generic;

    public class Queue<T> : IQueue<T>
    {
        private readonly LinkedList<T> list;

        public Queue()
        {
            this.list = new LinkedList<T>();
        }

        public int Count
        {
            get
            {
                return this.list.Count;
            }
        }

        public void Enqueue(T value)
        {
            this.list.AddLast(value);
        }

        public T Dequeue()
        {
            var value = this.list.First;
            this.list.RemoveFirst();
            return value;
        }

        public T Peek()
        {
            return this.list.First;
        }

        public bool IsEmpty()
        {
            return this.Count == 0;
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this.list.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        IEnumerator<T> IEnumerable<T>.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
}
